modules/files.nix

```diff
@@ -54,9 +64,40 @@ let
             # This means our `source` definition has the same priority as `text`.
             lib.mkDerivedConfig options.text (pkgs.writeText derivationName)
           );
+          finalSource =
+            # Byte compile lua files if performance.byteCompileLua option is enabled
+            if
+              lib.hasSuffix ".lua" config.target
+              && topConfig.performance.byteCompileLua.enable
+              && topConfig.performance.byteCompileLua.configs
+            then
+              if lib.isDerivation config.source then
+                # Source is a derivation, add byteCompileLuaHook to it
+                (config.source.overrideAttrs (
+                  prev:
+                  {
+                    nativeBuildInputs = prev.nativeBuildInputs or [ ] ++ [ helpers.byteCompileLuaHook ];
+                  }
+                  // lib.optionalAttrs (prev ? buildCommand) {
+                    buildCommand = ''
+                      ${prev.buildCommand}
+                      runHook postFixup
+                    '';
+                  }
+                ))
+              else
+                # Source is a path or string
+                helpers.writeByteCompiledLua derivationName (builtins.readFile config.source)
```

Comment:
This will lead to IFD if the `source` is a path within a derivation, or a stringified derivation.

To avoid this we'd need a builder function that accepts a path (instead of string content) and doesn't read the path's content during evaluation (only during its build)